CRM/Civicase/Hook/SelectWhereClause/RestrictSearchToAccessibleCaseCategories.php

```diff
@@ -0,0 +1,56 @@
+<?php
+
+use CRM_Civicase_Helper_CaseCategory as CaseCategoryHelper;
+
+/**
+ * Restrict search to accessible case categories if case entity is involved.
+ */
+class CRM_Civicase_Hook_SelectWhereClause_RestrictSearchToAccessibleCaseCategories {
+
+  /**
+   * Restrict search to accessible case categories if case entity is involved.
+   *
+   * @param string $entity
+   *   Entity name.
+   * @param array $clauses
+   *   Where clauses.
+   */
+  public function run($entity, array &$clauses) {
+
+    if (!$this->shouldRun($entity)) {
+      return;
+    }
+    $accessibleCaseCategories = CaseCategoryHelper::getAccessibleCaseTypeCategories();
+    $params = (json_decode($_REQUEST['json'], TRUE));
+
+    if (
+      isset($params[0][2]['case_type_id.case_type_category']) &&
+      (
+        (in_array(ucfirst($params[0][2]['case_type_id.case_type_category']), $accessibleCaseCategories)) ||
+        in_array($params[0][2]['case_type_id.case_type_category'], array_keys($accessibleCaseCategories))
+      )) {
+      return;
+    }
+    elseif (isset($params[0][2]['case_type_id.case_type_category'])) {
+      $clauses['case_type_id'] = 'IS NULL';
+    }
+    else {
+      $accessibleCaseCategoriesIds = implode(',', array_keys($accessibleCaseCategories));
+      $clauses['case_type_id'][] = "IN (SELECT id FROM civicrm_case_type WHERE case_type_category in ($accessibleCaseCategoriesIds))";
+    }
+  }
+
+  /**
+   * Whether the hook should run or not.
```

Comment:
Can we change this description to `Should Run when the entity is Case` or something more descriptive.